sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuTransitionOverrides.scala

```diff
@@ -55,21 +55,26 @@ class GpuTransitionOverrides extends Rule[SparkPlan] {
       GpuRowToColumnarExec(optimizeAdaptiveTransitions(r2c.child, Some(r2c)), goal)
 
     case ColumnarToRowExec(GpuBringBackToHost(
-      GpuCoalesceBatches(e: GpuShuffleExchangeExecBase, _))) if parent.isEmpty =>
-      // If this coalesced exchange has no parent it means that this plan is being created
-      // as a query stage when AQE is on. Because we must return an operator that implements
-      // ShuffleExchangeLike we need to remove the GpuCoalesceBatches operator here and
-      // re-insert it around the GpuCustomShuffleReaderExec in the parent query stage when that
-      // is created. Another option would be to create a custom operator that combines
-      // GpuCoalesceBatches and GpuShuffleExchangeExec and return that here instead. See
-      // https://github.com/NVIDIA/spark-rapids/issues/719 for more information.
-      optimizeAdaptiveTransitions(e, Some(plan))
+        GpuCoalesceBatches(e: GpuShuffleExchangeExecBase, _))) if parent.isEmpty =>
+      // new query stage
+      e.withNewChildren(e.children.map(c => optimizeAdaptiveTransitions(c, Some(e))))
```

Comment:
Yes, that is correct. Our existing rules translate the final step of the plan (the case where there is no parent operator) so that the data is bought back onto the host and returned as rows. However, when planning a new query stage with AQE, each shuffle ends up being a standalone query with no parent, so these rules don't make sense for that case and we remove them here. This is why, in 0.1,  AQE was so slow because we had unnecessary transitions around each exchange.

When we return the plan for a new query stage, the operator must either implement `BroadcastExchangeLike` or `ShuffleExchangeLike` so we can't return a coalesce operator here and we need to insert the coalesce when reading the results from a materialized query stage.

I will improve the documentation here.